pywren/jobrunner.py

```diff
@@ -1,33 +1,88 @@
 from tblib import pickling_support
 
 pickling_support.install()
 
+# def s3_url_parse(url):
+#     if url[:5] != "s3://":
+#         raise Exception("improperly formatted s3 url: {}".format(url))
+#     bucket, key = url[5:].split("/", 1)
+#     return bucket, key
+
 def b64str_to_bytes(str_data):
     str_ascii = str_data.encode('ascii')
     byte_data = base64.b64decode(str_ascii)
     return byte_data
 
 try:
-    func_filename = sys.argv[1]
-    data_filename = sys.argv[2]
-    out_filename = sys.argv[3]
+    jobrunner_config_filename = sys.argv[1]
+    out_filename = sys.argv[2]
+
+
     # initial output file in case job fails
     pickle.dump({'result' : None,
                  'success' : False},
                 open(out_filename, 'wb'), -1)
+    jobrunner_config = json.load(open(jobrunner_config_filename,
+                                      'r'))
+
+
+    # FIXME someday switch to storage handler
+    # download the func data into memory
+    s3_client = boto3.client("s3")
+    func_bucket, func_key = jobrunner_config['func_bucket'], jobrunner_config['func_key']
+    data_bucket, data_key = jobrunner_config['data_bucket'], jobrunner_config['data_key']
+
+    data_byte_range = jobrunner_config['data_byte_range']
+    func_obj_stream = s3_client.get_object(Bucket=func_bucket, Key=func_key)
+    loaded_func_all = pickle.loads(func_obj_stream['Body'].read())
+
+    # save modules, before we unpickle actual function
+    PYTHON_MODULE_PATH = jobrunner_config['python_module_path']
+    for m_filename, m_data in loaded_func_all['module_data'].items():
+        m_path = os.path.dirname(m_filename)
+
+        if len(m_path) > 0 and m_path[0] == "/":
+            m_path = m_path[1:]
+        to_make = os.path.join(PYTHON_MODULE_PATH, m_path)
+        try:
+            os.makedirs(to_make)
+        except OSError as e:
+            if e.errno == 17:
+                pass
+            else:
+                raise e
+        full_filename = os.path.join(to_make, os.path.basename(m_filename))
+        #print "creating", full_filename
+        with open(full_filename, 'wb') as fid:
+            fid.write(b64str_to_bytes(m_data))
+
+    # logger.info("Finished writing {} module files".format(len(d['module_data'])))
+    # logger.debug(subprocess.check_output("find {}".format(PYTHON_MODULE_PATH), shell=True))
+    # logger.debug(subprocess.check_output("find {}".format(os.getcwd()), shell=True))
+
+
+    # now unpickle function; it will expect modules to be there
+    loaded_func = pickle.loads(loaded_func_all['func'])
+
+    extra_get_args = {}
+    if data_byte_range is not None:
+        range_str = 'bytes={}-{}'.format(*data_byte_range)
+        extra_get_args['Range'] = range_str
+    data_obj_stream = s3_client.get_object(Bucket=data_bucket,
+                                           Key=data_key, **extra_get_args)
+    # FIXME make this streaming
+    loaded_data = pickle.loads(data_obj_stream['Body'].read())
```

Comment:
why is this not streaming right now ? you mean you want to stream the data into the function ?

Reply:
We're calling `.read()` which reads the entire object into memory as a string, and then `loads` which unpickles the string. The "right" thing to do would be pass the `data_object_stream['Body']` to `pickle.load` which wants a file-like object. The problem is that the streaming file-like object boto provides doesn't actually provide _enough_ of the file object API for pickle to use it. We could have a wrapper (like we do for the call to `tarfile`) but even that one is insufficient. 

So for N bytes of input data this means that we briefly have 2N. That said, the temporary string created is easy for GC to get rid of, and this is currently an order of magnitude better than what we had been storing in /tmp/